modules/homa-lite/src/lib.rs

```diff
@@ -293,29 +360,504 @@ pub mod module {
+							&who,
+							liquid_remaining.checked_sub(request_amount).expect("GE checked; qed"),
+						),
+						Ordering::Less => {
+							T::Currency::unreserve(
+								T::LiquidCurrencyId::get(),
+								&who,
+								request_amount.checked_sub(liquid_remaining).expect("LE checked; qed"),
+							);
+							Ok(())
+						}
+						_ => Ok(()),
+					}?;
+
+					// Insert/replace the new redeem request into storage.
+					RedeemRequests::<T>::insert(&who, (liquid_remaining, additional_fee));
+
+					Self::deposit_event(Event::<T>::RedeemRequested(who, liquid_remaining, additional_fee));
+				}
+			}
+			Ok(())
+		}
+
+		/// Request staking currencies to be unbonded from the Relaychain.
+		///
+		/// Requires `T::GovernanceOrigin`
+		///
+		/// Parameters:
+		/// - `staking_amount`: The amount of staking currency to be unbonded.
+		/// - `unbond_block`: The relaychain block number to unbond.
+		#[pallet::weight(< T as Config >::WeightInfo::schedule_unbond())]
+		#[transactional]
+		pub fn schedule_unbond(
+			origin: OriginFor<T>,
+			staking_amount: Balance,
+			unbond_block: RelaychainBlockNumberOf<T>,
+		) -> DispatchResult {
+			T::GovernanceOrigin::ensure_origin(origin)?;
+
+			ScheduledUnbond::<T>::append((staking_amount, unbond_block));
+
+			Self::deposit_event(Event::<T>::ScheduledUnbondAdded(staking_amount, unbond_block));
+			Ok(())
+		}
+
+		/// Replace the current storage for `ScheduledUnbond`.
+		/// This should only be used to correct mistaken call of schedule_unbond or if something
+		/// unexpected happened on relaychain.
+		///
+		/// Requires `T::GovernanceOrigin`
+		///
+		/// Parameters:
+		/// - `new_unbonds`: The new ScheduledUnbond storage to replace the currrent storage.
+		#[pallet::weight(< T as Config >::WeightInfo::replace_schedule_unbond())]
+		#[transactional]
+		pub fn replace_schedule_unbond(
+			origin: OriginFor<T>,
+			new_unbonds: Vec<(Balance, RelaychainBlockNumberOf<T>)>,
+		) -> DispatchResult {
+			T::GovernanceOrigin::ensure_origin(origin)?;
+
+			ScheduledUnbond::<T>::put(new_unbonds);
+
+			Self::deposit_event(Event::<T>::ScheduledUnbondReplaced);
+
+			Ok(())
+		}
 	}
-}
 
-pub struct LiquidExchangeProvider<T>(sp_std::marker::PhantomData<T>);
-impl<T: Config> ExchangeRateProvider for LiquidExchangeProvider<T> {
-	fn get_exchange_rate() -> ExchangeRate {
-		Pallet::<T>::get_staking_exchange_rate()
-			.reciprocal()
-			.unwrap_or_default()
+	impl<T: Config> Pallet<T> {
+		/// Calculate the exchange rate between the Staking and Liquid currency.
+		/// returns Ratio(liquid : staking) = liquid_total_issuance / total_staking_amount
+		/// If the exchange rate cannot be calculated, T::DefaultExchangeRate is used
+		fn get_exchange_rate() -> Ratio {
+			let staking_total = Self::total_staking_currency();
+			let liquid_total = T::Currency::total_issuance(T::LiquidCurrencyId::get());
+			Ratio::checked_from_rational(liquid_total, staking_total).unwrap_or_else(T::DefaultExchangeRate::get)
```

Comment:
`.unwrap_or_else(T::DefaultExchangeRate::get)` isn't right. it needs to be `1 / DefaultExchangeRate`

should have more tests to make sure `get_exchange_rate` / `convert_liquid_to_staking`  / `convert_staking_to_liquid` give correct number on default case